### ISSUE TITLE
Default values ignored with null encoded types

### ISSUE THREAD

**Issue by jefffisher** [billvaglienti/ProtoGen#73]
When using a `default` attribute prior to a `Data` tag that has an `encodedType` of `null` causes the default tag to be ignored and sometimes also has potentially dangerous side effects. For example:

```
<Packet name="OrionExtHeadingData" ID="ORION_PKT_EXT_HEADING_DATA" parameterInterface="true">
    <Data name="extHeading" inMemoryType="float32" encodedType="signed16" max="pi"/>
    <Data name="noise" inMemoryType="float32" encodedType="unsigned16" max="2*pi" default="3"/>
    <Data name="headingInGimbalAxis" inMemoryType="bitfield1" default="0"/>
    <Data name="reserved" inMemoryType="null" encodedType="bitfield15" default="0"/>
    <Data name="pitch" inMemoryType="float32" default="0" encodedType="signed16" max="pi"/>
</Packet>
```

This results in the following code output:

```
OrionPublicPacket.h:

//! return the minimum encoded length for the OrionExtHeadingData packet
#define getOrionExtHeadingDataMinDataLength() (4)

//! return the maximum encoded length for the OrionExtHeadingData packet
#define getOrionExtHeadingDataMaxDataLength() (8)
```

```
OrionPublicPacket.c:

int decodeOrionExtHeadingDataPacket(const void* pkt, float* extHeading, float* noise, unsigned* headingInGimbalAxis, float* pitch)
{
    int byteindex = 0;
    const uint8_t* data = getOrionPublicPacketDataConst(pkt);
    int numBytes = getOrionPublicPacketSize(pkt);

    // Verify the packet identifier
    if(getOrionPublicPacketID(pkt) != getOrionExtHeadingDataPacketID())
        return 0;

    if(numBytes < getOrionExtHeadingDataMinDataLength())
        return 0;

    // this packet has default fields, make sure they are set
    *pitch = 0;

    // Range of extHeading is -3.141592653589793 to 3.141592653589793.
    *extHeading = float32ScaledFrom2SignedBeBytes(data, &byteindex, 1.0f/10430.06004058427f);

    // Range of noise is 0.0 to 6.283185307179586.
    *noise = float32ScaledFrom2UnsignedBeBytes(data, &byteindex, 0.0f, 1.0f/10430.21919552736f);

    (*headingInGimbalAxis) = (data[byteindex] >> 7);

    byteindex += 2; // close bit field

    if(byteindex + 2 > numBytes)
        return 1;

    // Range of pitch is -3.141592653589793 to 3.141592653589793.
    *pitch = float32ScaledFrom2SignedBeBytes(data, &byteindex, 1.0f/10430.06004058427f);

    return 1;
}
```

Note that in the case where the data is a bitfield, the decode function actually **reads beyond** the 4-byte minimum length, and the only default field is actually the last one. Setting `inMemoryType="unsigned16"` on `reserved` generates these (correct) outputs:

```
OrionPublicPacket.h:

//! return the minimum encoded length for the OrionExtHeadingData packet
#define getOrionExtHeadingDataMinDataLength() (2)

//! return the maximum encoded length for the OrionExtHeadingData packet
#define getOrionExtHeadingDataMaxDataLength() (8)
```

```
OrionPublicPacket.c:

int decodeOrionExtHeadingDataPacket(const void* pkt, float* extHeading, float* noise, unsigned* headingInGimbalAxis, uint16_t* reserved, float* pitch)
{
    unsigned int tempbitfield = 0;
    int byteindex = 0;
    const uint8_t* data = getOrionPublicPacketDataConst(pkt);
    int numBytes = getOrionPublicPacketSize(pkt);

    // Verify the packet identifier
    if(getOrionPublicPacketID(pkt) != getOrionExtHeadingDataPacketID())
        return 0;

    if(numBytes < getOrionExtHeadingDataMinDataLength())
        return 0;

    // this packet has default fields, make sure they are set
    *noise = 3;
    *headingInGimbalAxis = 0;
    *reserved = 0;
    *pitch = 0;

    // Range of extHeading is -3.141592653589793 to 3.141592653589793.
    *extHeading = float32ScaledFrom2SignedBeBytes(data, &byteindex, 1.0f/10430.06004058427f);

    if(byteindex + 2 > numBytes)
        return 1;

    // Range of noise is 0.0 to 6.283185307179586.
    *noise = float32ScaledFrom2UnsignedBeBytes(data, &byteindex, 0.0f, 1.0f/10430.21919552736f);

    if(byteindex + 1 > numBytes)
        return 1;

    (*headingInGimbalAxis) = (data[byteindex] >> 7);

    if(byteindex + 2 > numBytes)
        return 1;

    tempbitfield = (data[byteindex] & 0x7F);

    tempbitfield <<= 8;
    tempbitfield |= data[byteindex + 1];

    (*reserved) = tempbitfield;
    byteindex += 2; // close bit field

    if(byteindex + 2 > numBytes)
        return 1;

    // Range of pitch is -3.141592653589793 to 3.141592653589793.
    *pitch = float32ScaledFrom2SignedBeBytes(data, &byteindex, 1.0f/10430.06004058427f);

    return 1;
}
```

**Comment by jefffisher** [comment]
Side note: I never would've noticed this without the handy new warning messages!

```
OrionPublicProtocol.xml:978:0: warning: OrionPublic:OrionExtHeadingData:noise: default value ignored, field is followed by non-default
OrionPublicProtocol.xml:979:0: warning: OrionPublic:OrionExtHeadingData:headingInGimbalAxis: default value ignored, field is followed by non-default
```

**Comment by billvaglienti** [comment]
This was actually quite subtle. It's an interesting problem: generally you cannot follow a default field with a non-default field. However if that non-default field is not in memory (inMemoryType = "null") then you can do that. However ProtoGen was not correctly allowing that case. I think this is fixed by 2.14.e, but there could still be subtle cases that I missed.
